### ISSUE TITLE
Orientation of Raster elements.

### ISSUE THREAD

**Issue by basnijholt** [ioam/holoviews#155]
I think ``Image`` and ``Raster`` could have a more natural orientation.

Notice the following behaviour, in which I want to depict the maximum value of a gaussian.
```python 
gaussian = lambda x, y, x0=1, y0=0.5: np.exp(-0.5 * ((x-x0)**2 + (y-y0)**2))
xs = np.linspace(-1, 1, 100)
ys = np.linspace(-1, 1, 150)
data = [gaussian(x, y) for x, y in product(xs, ys)]
data = np.reshape(data, (len(xs), len(ys)))

xmax, ymax = np.unravel_index(data.argmax(), data.shape)

im = hv.Raster(cone_max, kdims=['x', 'y'])
im + im * hv.Scatter([[xmax, ymax]])
```

```
print("x is {}, y is {}".format(xmax, ymax))
x is 99, y is 112
```

![download 4](https://cloud.githubusercontent.com/assets/6897215/8182885/0ca98824-1434-11e5-940a-3b7704ffa85a.png)

We start with a very default convention, getting the data as ``(x, y)``, so I can call if by ``data[x, y]``.
Now I plot it and suddenly the axis are swapped.

Also the y-axis for the ``Overlay`` of the ``Scatter`` and ``Raster`` is suddenly reversed.

Now I can get the orientation of the ``Raster`` right by using ``np.rot90``.

```python
im = hv.Raster(np.rot90(cone_max), kdims=['x', 'y'])
im + im * hv.Scatter([[xmax, ymax]])
```
![download 5](https://cloud.githubusercontent.com/assets/6897215/8182954/aa2443f0-1434-11e5-915b-1c31d7faf44f.png)

Now here the picture on the left is how I would like to see it, but then with the y-axis reversed!

Take the following example for more (for me) unintuitive behaviour:
```python
im = hv.Raster(np.rot90(cone_max), kdims=['x', 'y'])
im2 = hv.Image(np.rot90(cone_max), kdims=['x', 'y'], bounds=(0, 0, 100, 150))
im + im * hv.Scatter([[xmax, ymax]]) + im2 + im2 * hv.Scatter([[xmax, ymax]])
```

![download 6](https://cloud.githubusercontent.com/assets/6897215/8183040/5ba1bbd0-1435-11e5-811d-739413b72264.png)

Finally the ``Image`` on the far right is what I wanted the first time by doing ``hv.Raster(data)``.

This is all very confusing and could be much better in my opinion.

**Comment by jlstevens** [comment]
> Also the y-axis for the Overlay of the Scatter and Raster is suddenly reversed.

This bit is definitely a bug that should get fixed soon. I'll have to spend a little while thinking about the rest of your suggestion before forming an opinion.

**Comment by basnijholt** [comment]
There is of course also people that use ``np.meshgrid`` to evaluate functions on a grid.

Then there are two options:
* ``np.meshgrid(xs, ys, indexing='xy')``, this has the same awkward behaviour as Matlab, it transposes the grid.
* ``np.meshgrid(xs, ys, indexing='ij')``, this one is the same as in my original example.

I will plot the gaussian again with the x-average on zero and y-average on 0.5.
```
gaussian = lambda x, y, x0=0, y0=0.5: np.exp(-0.5 * ((x-x0)**2 + (y-y0)**2))

xs = np.linspace(-1, 1, 100)
ys = np.linspace(-1, 1, 150)

X, Y = np.meshgrid(xs, ys, indexing='xy')
data1 = gaussian(X, Y)
print(data1.shape)
>>> (150, 100)



X, Y = np.meshgrid(xs, ys, indexing='ij')
data2 = gaussian(X, Y)
print(data2.shape)
(data1 == gaussian(X, Y).T).all()
>>> (100, 150), True
```

To see the picture that you would imagine:
```
(hv.Image(np.flipud(data1), bounds=(-1,-1,1,1), label="np.meshgrid(xs, ys, indexing='xy')")
+ hv.Image(np.rot90(data2), bounds=(-1,-1,1,1), label="np.meshgrid(xs, ys, indexing='ij')"))
```
![download 7](https://cloud.githubusercontent.com/assets/6897215/8185933/6c2f3542-1447-11e5-8866-cc208f3ad206.png)

In both cases of the use of ``meshgrid`` you have to rotate or flip the data.

**Comment by jlstevens** [comment]
Ok, I have to admit I am not quite following the problem here but I had a look to see if ``Raster`` behaves the way I expect i.e like the default matplotlib ``imshow`` command:

```python
from matplotlib import pyplot as plt
import numpy as np
%matplotlib inline
```

Then:

```python
data = np.random.rand(5,3)
plt.imshow(data)
```
![image](https://cloud.githubusercontent.com/assets/890576/8185941/1409f0bc-143f-11e5-86e7-940896089889.png)

```python
import holoviews as hv
raster = hv.Raster(data)
raster
```
![image](https://cloud.githubusercontent.com/assets/890576/8185969/47709136-143f-11e5-921e-f99c09aac065.png)

```python
data[0,0] == raster.data[0,0] == raster[0,0]
True
```

This looks like the expected behavior to me! Maybe the issue is that you are used to using  ``pcolormesh`` or similar where you have explicit ``x`` and ``y`` arrays? I see you've posted something while I wrote this so I will comment now and then catch up!

**Comment by jbednar** [comment]
I only  have a minute now, so I can't work through the examples in detail, but some observations:

 * The way things are *supposed* to work in HoloViews are shown at http://ioam.github.io/holoviews/Tutorials/Continuous_Coordinates.html . I.e., both Raster and Image are meant to have a visual display that corresponds exactly to the way Numpy prints an array on screen.  In both cases, the underlying Numpy array should match how the data shows up on your monitor.  The difference between the two is meant to be that a Raster is indexed the way a Numpy array is, with zero in the upper left and rows and columns going out from there, while an Image is indexed in the normal Cartesian coordinates.  Both of these have their uses.  There may be some other convention that's useful, but if so I don't know of it!

* Scientific packages and graphics packages have very little consistency in how they handle samples on a grid.  E.g. many graphics packages treat the upper left pixel as if it starts at 0,0, but in terms of human perception this is clearly wrong -- the pixel is actually centered at d/2,d/2, where d is the width of one pixel.  Thus if you take some data generated from meshgrid, sampling some function at regular intervals starting at e.g. 0 and then try to visualize it, your data can easily end up visualized as being at a position off from its actual position by d/2.  Hopefully the above link makes it clear what HoloViews is assuming -- it visualizes each sample as being the center of a pixel, because that's the way a human will perceive it, and you should make sure that the data you give HoloViews is set up to match that.  

For small enough pixels, of course, the difference is negligible, but there's no reason not to do it right!
Hope that makes sense...

**Comment by jlstevens** [comment]
> Also the y-axis for the Overlay of the Scatter and Raster is suddenly reversed.

This bug at least should be fixed for version 1.2.5

**Comment by philippjfr** [comment]
I've now addressed the bug where Raster Elements were inverted when overlaid but you did raise some other issues. As far as I can tell from your examples the only thing you need to do different is to invert the Y and X variables you pass to the Gaussian function, i.e.:

```
gaussian = lambda x, y, x0=0, y0=0.5: np.exp(-0.5 * ((x-x0)**2 + (y-y0)**2))

xs = np.linspace(-1, 1, 100)
ys = np.linspace(-1, 1, 150)

X, Y = np.meshgrid(xs, ys)
data1 = gaussian(Y, X)
hv.Raster(data1)
```

And then you can invert the Raster y_axis with ``invert_yaxis=True``. On that basis I'll close this issue now.
